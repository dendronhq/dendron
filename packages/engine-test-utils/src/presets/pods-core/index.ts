import JSON from "./json";
import MARKDOWN from "./markdown";

export const PODS_CORE = {
  JSON,
  MARKDOWN,
};

export const PODS_PRESETS = [{ name: "json", presets: JSON }];
