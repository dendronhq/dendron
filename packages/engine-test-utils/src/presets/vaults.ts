import { DVault } from "@dendronhq/common-all";

export const VAULTS = {
  MULTI_VAULT_WITH_THREE_VAULTS: () => {
    return [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ] as DVault[];
  },
};
