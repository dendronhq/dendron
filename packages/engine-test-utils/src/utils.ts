import {
  CONSTANTS,
  DendronConfig,
  VaultUtils,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings,
} from "@dendronhq/common-all";
import { readYAML } from "@dendronhq/common-server";
import { AssertUtils } from "@dendronhq/common-test-utils";
import { DConfig, Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

const getWorkspaceFolders = (wsRoot: string) => {
  const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME);
  const settings = fs.readJSONSync(wsPath) as WorkspaceSettings;
  return _.toArray(settings.folders);
};

export function checkVaults(opts: WorkspaceOpts, expect: any) {
  const { wsRoot, vaults } = opts;
  const configPath = DConfig.configPath(opts.wsRoot);
  const config = readYAML(configPath) as DendronConfig;
  expect(_.sortBy(config.vaults, ["fsPath", "workspace"])).toEqual(
    _.sortBy(vaults, ["fsPath", "workspace"])
  );
  const wsFolders = getWorkspaceFolders(wsRoot);
  expect(wsFolders).toEqual(
    vaults.map((ent) => {
      const out: WorkspaceFolderRaw = { path: VaultUtils.getRelPath(ent) };
      if (ent.name) {
        out.name = ent.name;
      }
      return out;
    })
  );
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }

  /** Creates a "bare" git repository, to be used as the remote for a workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteCreate(remoteDir: string) {
    const git = new Git({ localUrl: remoteDir, bare: true });
    await git.init();
  }

  /** Adds a bare repository created with `createRemote` as the remote for the workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteAdd(wsRoot: string, remoteDir: string) {
    const git = new Git({ localUrl: wsRoot, remoteUrl: remoteDir });
    await git.remoteAdd();
    // Need to push to be able to set up remote tracking branch
    await git.push({ remote: "origin", branch: await git.getCurrentBranch() });
  }

  /** Set up a workspace with a remote, intended to be used when testing pull or push functionality.
   *
   * @param wsRoot Directory where the workspace will be stored.
   * @param remoteDir Directory where the remote will be stored. The workspace will pull and push to this remote.
   */
  static async createRepoForRemoteWorkspace(wsRoot: string, remoteDir: string) {
    await this.createRepoForWorkspace(wsRoot);
    await this.remoteCreate(remoteDir);
    await this.remoteAdd(wsRoot, remoteDir);
  }

  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
