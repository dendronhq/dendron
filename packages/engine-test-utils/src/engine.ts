import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
