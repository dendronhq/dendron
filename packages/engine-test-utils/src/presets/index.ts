import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
