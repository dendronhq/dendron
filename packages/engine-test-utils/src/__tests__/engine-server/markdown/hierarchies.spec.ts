import {
  ConfigService,
  DendronConfig,
  NoteDictsUtils,
  NoteProps,
  URI,
} from "@dendronhq/common-all";
import {
  NoteTestUtilsV4,
  TestPresetEntryV4,
} from "@dendronhq/common-test-utils";
import {
  DendronASTDest,
  getParsingDependencyDicts,
  MDUtilsV5,
  ProcDataFullOptsV5,
} from "@dendronhq/unified";
import { runEngineTestV5 } from "../../../engine";
import { ENGINE_HOOKS } from "../../../presets";
import { checkNotInVFile, checkVFile, createProcTests } from "./utils";

function procDendronForPublish(
  opts: Omit<ProcDataFullOptsV5, "dest"> & {
    noteIndex: NoteProps;
  }
) {
  const { config, fname, vault, noteToRender, noteCacheForRenderDict } = opts;
  const proc = MDUtilsV5.procRehypeFull({
    noteToRender,
    config,
    fname,
    vault,
    noteCacheForRenderDict,
  });
  return proc;
}

describe("hierarchies", () => {
  const BASIC_TEXT = "[Ch1](foo.ch1.html)";
  const BASIC = createProcTests({
    name: "BASIC",
    setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
      const config = (
        await ConfigService.instance().readConfig(URI.file(wsRoot))
      )._unsafeUnwrap();
      config.publishing.enableHierarchyDisplay = true;
      const noteToRender = (
        await engine.findNotes({ fname: "foo", vault: vaults[0] })
      )[0];
      const noteCacheForRenderDict = await getParsingDependencyDicts(
        noteToRender,
        engine,
        config,
        vaults
      );
      NoteDictsUtils.add(
        (await engine.findNotes({ fname: "foo.ch1", vault: vaults[0] }))[0],
        noteCacheForRenderDict
      );
      if (extra.dest !== DendronASTDest.HTML) {
        const proc = MDUtilsV5.procRemarkFull({
          noteToRender,
          noteCacheForRenderDict,
          fname: "foo",
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      } else {
        const proc = procDendronForPublish({
          noteToRender,
          noteCacheForRenderDict,
          fname: "foo",
          noteIndex: (await engine.getNote("foo")).data!,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      }
    },
    verifyFuncDict: {
      [DendronASTDest.MD_DENDRON]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(resp, BASIC_TEXT);
      },
      [DendronASTDest.HTML]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(resp, "Children");
      },
    },
    preSetupHook: ENGINE_HOOKS.setupBasic,
  });

  const NO_HIERARCHY = createProcTests({
    name: "NO_HIERARCHY",
    setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
      const rawConfig = (
        await ConfigService.instance().readRaw(URI.file(wsRoot))
      )._unsafeUnwrap() as DendronConfig;
      const config: DendronConfig = {
        ...rawConfig,
        publishing: {
          ...rawConfig.publishing,
          enableHierarchyDisplay: false,
        },
      };
      if (extra.dest !== DendronASTDest.HTML) {
        const proc = MDUtilsV5.procRemarkFull({
          noteToRender: (await engine.getNote("foo")).data!,
          fname: "foo",
          config,
          dest: extra.dest,
          vault: vaults[0],
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      } else {
        const proc = procDendronForPublish({
          noteToRender: (await engine.getNote("foo")).data!,
          config,
          fname: "foo",
          noteIndex: (await engine.getNote("foo")).data!,
          vault: vaults[0],
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      }
    },
    verifyFuncDict: {
      [DendronASTDest.MD_DENDRON]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(resp, BASIC_TEXT);
      },
      [DendronASTDest.HTML]: async ({ extra }) => {
        const { resp } = extra;
        await checkNotInVFile(resp, "Children");
      },
    },
    preSetupHook: ENGINE_HOOKS.setupBasic,
  });

  const NO_HIERARCHY_VIA_FM = createProcTests({
    name: "NO_HIERARCHY_VIA_FM",
    setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
      const config = (
        await ConfigService.instance().readConfig(URI.file(wsRoot))
      )._unsafeUnwrap();
      if (extra.dest !== DendronASTDest.HTML) {
        const proc = MDUtilsV5.procRemarkFull({
          noteToRender: (await engine.getNote("foo")).data!,
          fname: "foo",
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      } else {
        const proc = procDendronForPublish({
          noteToRender: (await engine.getNote("foo")).data!,
          fname: "foo",
          noteIndex: (await engine.getNote("foo")).data!,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      }
    },
    verifyFuncDict: {
      [DendronASTDest.MD_DENDRON]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(resp, BASIC_TEXT);
      },
      [DendronASTDest.HTML]: async ({ extra }) => {
        const { resp } = extra;
        await checkNotInVFile(resp, "Children");
      },
    },
    preSetupHook: async (opts) => {
      await ENGINE_HOOKS.setupBasic(opts);
      await NoteTestUtilsV4.modifyNoteByPath(
        { wsRoot: opts.wsRoot, vault: opts.vaults[0], fname: "foo" },
        (note: NoteProps) => {
          note.custom.hierarchyDisplay = false;
          return note;
        }
      );
    },
  });

  const DIFF_HIERARCHY_TITLE = createProcTests({
    name: "DIFF_HIERARCHY_TITLE",
    setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
      const rawConfig = (
        await ConfigService.instance().readRaw(URI.file(wsRoot))
      )._unsafeUnwrap() as DendronConfig;
      const config: DendronConfig = {
        ...rawConfig,
        publishing: {
          ...rawConfig.publishing,
          hierarchyDisplayTitle: "Better Children",
        },
      };
      const noteToRender = (
        await engine.findNotes({ fname: "foo", vault: vaults[0] })
      )[0];
      const noteCacheForRenderDict = await getParsingDependencyDicts(
        noteToRender,
        engine,
        config,
        vaults
      );
      if (extra.dest !== DendronASTDest.HTML) {
        const proc = MDUtilsV5.procRemarkFull({
          noteToRender,
          noteCacheForRenderDict,
          fname: "foo",
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      } else {
        const proc = procDendronForPublish({
          noteToRender,
          noteCacheForRenderDict,
          fname: "foo",
          noteIndex: (await engine.getNote("foo")).data!,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      }
    },
    verifyFuncDict: {
      [DendronASTDest.HTML]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(resp, "Better Children");
      },
    },
    preSetupHook: ENGINE_HOOKS.setupBasic,
  });

  const SKIP_LEVELS = createProcTests({
    name: "SKIP_LEVELS",
    setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
      const config = (
        await ConfigService.instance().readConfig(URI.file(wsRoot))
      )._unsafeUnwrap();
      const noteToRender = (
        await engine.findNotes({ fname: "daily", vault: vaults[0] })
      )[0];
      const noteCacheForRenderDict = await getParsingDependencyDicts(
        noteToRender,
        engine,
        config,
        vaults
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal.2020",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal.2020.07",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal.2020.07.01",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal.2020.07.01.one",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal.2020.07.05",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      NoteDictsUtils.add(
        (
          await engine.findNotes({
            fname: "daily.journal.2020.07.05.two",
            vault: vaults[0],
          })
        )[0],
        noteCacheForRenderDict
      );
      if (extra.dest !== DendronASTDest.HTML) {
        const proc = MDUtilsV5.procRemarkFull({
          noteToRender,
          noteCacheForRenderDict,
          fname: "daily",
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      } else {
        const proc = procDendronForPublish({
          noteToRender,
          noteCacheForRenderDict,
          fname: "daily",
          noteIndex: (await engine.getNote("daily")).data!,
          vault: vaults[0],
          config,
        });
        const resp = await proc.process(BASIC_TEXT);
        return { resp };
      }
    },
    verifyFuncDict: {
      [DendronASTDest.MD_DENDRON]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(resp, BASIC_TEXT);
      },
      [DendronASTDest.HTML]: async ({ extra }) => {
        const { resp } = extra;
        await checkVFile(
          resp,
          "daily.journal.2020.07.01.one",
          "daily.journal.2020.07.05.two"
        );
      },
    },
    preSetupHook: async (opts) => {
      await ENGINE_HOOKS.setupJournals(opts);
      await NoteTestUtilsV4.modifyNoteByPath(
        { fname: "daily", vault: opts.vaults[0], wsRoot: opts.wsRoot },
        (note) => {
          note.custom.skipLevels = 4;
          return note;
        }
      );
    },
  });

  const ALL_TEST_CASES = [
    ...NO_HIERARCHY,
    ...NO_HIERARCHY_VIA_FM,
    ...BASIC,
    ...DIFF_HIERARCHY_TITLE,
    ...SKIP_LEVELS,
  ];

  test.each(
    ALL_TEST_CASES.map((ent) => [`${ent.dest}: ${ent.name}`, ent.testCase])
    // @ts-ignore
  )("%p", async (_key, testCase: TestPresetEntryV4) => {
    await runEngineTestV5(testCase.testFunc, {
      expect,
      preSetupHook: testCase.preSetupHook,
    });
  });
});
