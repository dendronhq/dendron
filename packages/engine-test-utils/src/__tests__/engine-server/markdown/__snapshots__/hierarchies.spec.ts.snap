// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hierarchies "HTML: BASIC" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo</h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"foo.ch1\\">Ch1</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "HTML: DIFF_HIERARCHY_TITLE" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo</h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>
<hr>
<strong>Better Children</strong>
<ol>
<li><a href=\\"foo.ch1\\">Ch1</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "HTML: NO_HIERARCHY" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo</h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "HTML: NO_HIERARCHY_VIA_FM" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo</h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "HTML: SKIP_LEVELS" 1`] = `
VFile {
  "contents": "<h1 id=\\"daily\\">Daily</h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"daily.journal.2020.07.01.one\\">One</a></li>
<li><a href=\\"daily.journal.2020.07.05.two\\">Two</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "MD_DENDRON: BASIC" 1`] = `
VFile {
  "contents": "# Foo

[Ch1](foo.ch1.html)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "MD_DENDRON: NO_HIERARCHY" 1`] = `
VFile {
  "contents": "# Foo

[Ch1](foo.ch1.html)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "MD_DENDRON: NO_HIERARCHY_VIA_FM" 1`] = `
VFile {
  "contents": "# Foo

[Ch1](foo.ch1.html)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`hierarchies "MD_DENDRON: SKIP_LEVELS" 1`] = `
VFile {
  "contents": "# Daily

[Ch1](foo.ch1.html)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
